### PR TITLE
Hyperboloid

### PR DESCRIPTION
**Pull request by rrkarim**
There are some numerical issues, and I'm starting to think that there is no easy solution for it at this point. Maybe something like this [paper](https://papers.nips.cc/paper/8476-numerically-accurate-hyperbolic-embeddings-using-tiling-based-models.pdf) should be investigated more.
- [x] Add curvature
- [x] Add tests
- [x] Add an example with hgnn/attention [example with basic usage]
- [x] Add docs

Example with optimization should be added on the separate PR

### CODE REVIEW

**Comment by rrkarim** [comment]
@ferrine fair, I'm working on it. It also should be tested in hgnn.

**Comment by rrkarim** [comment]
@ferrine can you please check it when you have some time, there are still some issues I see but I would like to see what your concerns are.

File: geoopt/manifolds/lorentz/math.py [hunk 1]
```diff
@@ -0,0 +1,607 @@
...
+import torch as th
+
+MIN_NORM = 1e-15
+
+
+class Arcosh(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, x, eps):
+        z = th.sqrt(th.clamp_min(x.pow(2) - 1.0, MIN_NORM))
+        ctx.save_for_backward(z)
```
Comment (ferrine) [inline geoopt/manifolds/lorentz/math.py]:
```suggestion
        ctx.save_for_backward(z, eps)
```

**Comment by ferrine** [comment]
Can you use manifold in tests? It will give you better coverage

File: geoopt/manifolds/lorentz/__init__.py [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+"""
+
+
+class Lorentz(Manifold):
+    __doc__ = r"""{}
+    """
+
+    ndim = 1
+    reversible = False
+    name = "Hyperboloid"
```
Comment (ferrine) [inline geoopt/manifolds/lorentz/__init__.py]: Why not Lorenz?

File: geoopt/manifolds/lorentz/__init__.py [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+            reason = None
+        return ok, reason
+
+    def _check_vector_on_tangent(
+        self, x: torch.Tensor, u: torch.Tensor, *, atol=1e-5, rtol=1e-5
+    ) -> Tuple[bool, Optional[str]]:
+        inner_ = math.inner(u, x)
+        ok = torch.allclose(inner_, torch.zeros(1), atol=atol, rtol=rtol)
+        if not ok:
+            reason = f"Minkowski inner produt is not equal to zero"
```
Comment (ferrine) [inline geoopt/manifolds/lorentz/__init__.py]: f string is unnecessary here

File: geoopt/manifolds/lorentz/__init__.py [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+        return math.dist(x, y, k=self.k, keepdim=keepdim, dim=dim)
+
+    def norm(self, u: torch.Tensor, *, keepdim=False, dim=-1) -> torch.Tensor:
+        return math.norm(u, keepdim=keepdim, dim=dim)
+
+    def egrad2rgrad(self, x: torch.Tensor, u: torch.Tensor, *, dim=-1) -> torch.Tensor:
+        return math.egrad2rgrad(x, u, dim=dim)
+
+    def projx(self, x: torch.Tensor, dim=-1) -> torch.Tensor:
+        return math.project(x, k=self.k, dim=dim)
```
Comment (ferrine) [inline geoopt/manifolds/lorentz/__init__.py]: keepdim and dim arguments should be keyword only

File: geoopt/manifolds/lorentz/__init__.py [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+    """
+
+    ndim = 1
+    reversible = False
+    name = "Hyperboloid"
+    __scaling__ = Manifold.__scaling__.copy()
+
+    def __init__(self, k=1.0):
+        super().__init__()
+        self.register_buffer("k", torch.as_tensor(k, dtype=torch.get_default_dtype()))
```
Comment (ferrine) [inline geoopt/manifolds/lorentz/__init__.py]: Btw, you can make it learnable as done in Poincare.

File: geoopt/manifolds/lorentz/__init__.py [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+            zero point on the manifold
+        """
+        if not dtype:
+            dtype = self.k.dtype
+        if not device:
+            device = self.k.device
+
+        zero_point = torch.ones(*size, dtype=dtype, device=device)
+        d = zero_point.size(-1) - 1
+        zero_point = torch.cat(
```
Comment (ferrine) [inline geoopt/manifolds/lorentz/__init__.py]: You could do inplace updates with narrow

File: geoopt/manifolds/lorentz/math.py [hunk 7]
```diff
@@ -0,0 +1,607 @@
...
+        reduction dimension
+    eps : float
+        stability parameter for arcosh
+
+    Returns
+    -------
+    tensor
+        geodesic distance between :math:`x` and :math:`0`
+    """
+    if isinstance(k, float):
```
Comment (ferrine) [inline geoopt/manifolds/lorentz/math.py]: Let's not use this. Just tensors

File: geoopt/manifolds/lorentz/math.py [hunk 8]
```diff
@@ -0,0 +1,607 @@
+import torch.jit
+import torch as th
+
+MIN_NORM = 1e-15
+
+
+class Arcosh(torch.autograd.Function):
```
Comment (ferrine) [inline geoopt/manifolds/lorentz/math.py]: I no see little point in doing this. We could use torch script everywhere. It is well supported since 1.3.1. see K-stereographic for reference.

File: geoopt/manifolds/lorentz/math.py [hunk 9]
```diff
@@ -0,0 +1,607 @@
...
+        (zp.narrow(dim, 0, 1) * th.sqrt(k), zp.narrow(dim, 1, d) * 0.0), dim=dim
+    )
+
+    nom = inner(logmap(zp, y, k=k, dim=dim), v, keepdim=True)
+    denom = dist(zp, y, k=k, dim=dim, keepdim=True) ** 2
+    p = v - nom / denom * (logmap(zp, y, k=k, dim=dim) + logmap(y, zp, k=k, dim=dim))
+    return p
+
+
+def geodesic_unit(t, x, u, k=1.0):
```
Comment (ferrine) [inline geoopt/manifolds/lorentz/math.py]: No `dim` argument, k should also be keyword only. As for kw only, that is the only point to have a private implementation btw. Torch script doesn't support advanced signatures.

File: geoopt/manifolds/lorentz/__init__.py (hunk 5 above)
Comment (rrkarim) [reply geoopt/manifolds/lorentz/__init__.py]: You sure poincare has learnable params?

**Comment by ferrine** [comment]
It will allow for this
https://github.com/geoopt/geoopt/blob/c24046f25bce598b56e4fd5a4762e7826a0721e4/geoopt/manifolds/stereographic/manifold.py#L120

ср, 22 янв. 2020 г., 1:05 Rasul Karimov <notifications@github.com>:

File: geoopt/manifolds/lorentz/__init__.py [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+
+    def _check_point_on_manifold(
+        self, x: torch.Tensor, *, atol=1e-5, rtol=1e-5
+    ) -> Tuple[bool, Optional[str]]:
+        dn = x.size(0)
+        x = x ** 2
+        quad_form = -x[0] + x[1:].sum()
+        ok = torch.allclose(quad_form, -self.k, atol=atol, rtol=rtol)
+        if not ok:
+            reason = f"'x' minkowski quadratic form is not equal to {-self.k.item()}"
```
Comment (ferrine) [inline geoopt/manifolds/lorentz/__init__.py]: k may be not a scalar

File: geoopt/manifolds/lorentz/__init__.py [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+            the desired dtype
+        seed : int
+            ignored
+
+        Returns
+        -------
+        ManifoldTensor
+            zero point on the manifold
+        """
+        if not dtype:
```
Comment (ferrine) [inline geoopt/manifolds/lorentz/__init__.py]: Check if None

File: geoopt/manifolds/lorentz/math.py [hunk 12]
```diff
@@ -0,0 +1,578 @@
+import torch.jit
+import torch as th
```
Comment (ferrine) [inline geoopt/manifolds/lorentz/math.py]: Why th?  It is pronounced weirdly. `%s/ th./ torch./g`

File: geoopt/manifolds/lorentz/math.py [hunk 13]
```diff
@@ -0,0 +1,578 @@
...
+
+    Returns
+    -------
+    tensor
+        geodesic distance between :math:`x` and :math:`y`
+    """
+    return _dist(x, y, k=k, keepdim=keepdim, dim=dim, eps=eps)
+
+
+def _dist(x, y, k: th.Tensor, keepdim: bool = False, dim: int = -1, eps=1e-6):
```
Comment (ferrine) [inline geoopt/manifolds/lorentz/math.py]: Let's jit everything?

File: geoopt/manifolds/lorentz/math.py [hunk 14]
```diff
@@ -0,0 +1,578 @@
...
+
+    Returns
+    -------
+    tensor
+        projected vector on the manifold
+    """
+    return _project_u(x, v, k=k, dim=dim)
+
+
+def _project_u(x, v, k: th.Tensor, dim=-1):
```
Comment (ferrine) [inline geoopt/manifolds/lorentz/math.py]: Does it work without expand?

File: geoopt/manifolds/lorentz/math.py [hunk 15]
```diff
@@ -0,0 +1,578 @@
...
+
+
+def _parallel_transport(x, y, v, k, dim: int = -1):
+    nom = inner(logmap(x, y, k=k, dim=dim), v, keepdim=True)
+    denom = dist(x, y, k=k, dim=dim, keepdim=True) ** 2
+    p = v - nom / denom * (logmap(x, y, k=k, dim=dim) + logmap(y, x, k=k, dim=dim))
+    return p
+
+
+def parallel_transport0(y, v, *, k, dim=-1):
```
Comment (ferrine) [inline geoopt/manifolds/lorentz/math.py]: Parallel transport back to zero would be cool to

File: geoopt/manifolds/lorentz/__init__.py (hunk 10 above)
Comment (rrkarim) [reply geoopt/manifolds/lorentz/__init__.py]: I don't see the problem here.

**Comment by rrkarim** [comment]
@ferrine how do you solve the issue of coverage for traced ops?

File: geoopt/manifolds/lorentz/__init__.py (hunk 10 above)
Comment (rrkarim) [reply geoopt/manifolds/lorentz/__init__.py]: I see that now, you mean non constant curvature, ok.

**Comment by ferrine** [comment]
I didn't solve it. Ideally I would just run both PYTORCH_JIT=true and false. But it would fail with tensordot. API in torch script is magically different

File: geoopt/manifolds/lorentz/__init__.py [hunk 16]
```diff
@@ -0,0 +1,217 @@
...
+        if not torch.is_floating_point(k):
+            k = k.to(torch.get_default_dtype())
+        self.k = torch.nn.Parameter(k, requires_grad=learnable)
+
+    def _check_point_on_manifold(
+        self, x: torch.Tensor, *, atol=1e-5, rtol=1e-5
+    ) -> Tuple[bool, Optional[str]]:
+        dn = x.size(0)
+        x = x ** 2
+        quad_form = -x[0] + x[1:].sum()
```
Comment (ferrine) [inline geoopt/manifolds/lorentz/__init__.py]: this does not broadcast

Comment (ferrine) [reply geoopt/manifolds/lorentz/__init__.py]: also, please, add missing dim argument everywhere where it is not present

Comment (rrkarim) [reply geoopt/manifolds/lorentz/__init__.py]: I think you are looking at the old commit.

Comment (rrkarim) [reply geoopt/manifolds/lorentz/__init__.py]: These are fixed on the most recent commit.

File: geoopt/manifolds/lorentz/__init__.py [hunk 17]
```diff
@@ -0,0 +1,228 @@
...
+                "`device` does not match the projector `device`, set the `device` argument to None"
+            )
+        if dtype is not None and dtype != self.k.dtype:
+            raise ValueError(
+                "`dtype` does not match the projector `dtype`, set the `dtype` arguement to None"
+            )
+        tens = torch.randn(*size, device=self.k.device, dtype=self.k.dtype) * std + mean
+        tens /= tens.norm(dim=-1, keepdim=True)
+        return geoopt.ManifoldTensor(self.expmap0(tens), manifold=self)
+
```
Comment (ferrine) [inline geoopt/manifolds/lorentz/__init__.py]:
```suggestion
    random = random_normal
```

**Comment by rrkarim** [comment]
@ferrine I fixed the optimization issue, you can check for PR.

**Comment by rrkarim** [comment]
@ferrine I've fixed issues, I guess we can merge after your review. Maybe I should look at the docs, how can I launch docs on the local machine?

**Comment by rrkarim** [comment]
We should set xfail for `pytorch=1.2` tests. I guess 1.2 jit tracer doesn't support dtype tensor conversions.

**Comment by rrkarim** [comment]
> We should set xfail for `pytorch=1.2` tests. I guess 1.2 torch jit tracer doesn't support tensor dtype conversions. `pytorch>=1.3` tests are good.

ok, so in master branch 1.2 is removed.

File: test-output.xml [hunk 18]
```diff
@@ -1 +1 @@
...pe0]" time="0.003"></testcase><testcase classname="tests.test_lorentz_math" file="tests/test_lorentz_math.py" line="86" name="test_geodesic_segement_unit_property[39-dtype0]" time="0.004"></testcase><testcase classname="tests.test_lorentz_math" file="tests/test_lorentz_math.py" line="115" name="test_expmap0_logmap0[39-dtype0]" time="0.003"></testcase><testcase classname="tests.test_lorentz_math" file="tests/test_lorentz_math.py" line="127" name="test_parallel_transport0_preserves_inner_products[39-dtype0]" time="0.003"></testcase><testcase classname="tests.test_lorentz_math" file="tests/test_lorentz_math.py" line="151" name="test_parallel_transport0_back[39-dtype0]" time="0.004"></testcase><testcase classname="tests.test_lorentz_math" file="tests/test_lorentz_math.py" line="174" name="test_parallel_transport0_is_same_as_usual[39-dtype0]" time="0.003"></testcase><testcase classname="tests.test_lorentz_math" file="tests/test_lorentz_math.py" line="190" name="test_zero_point_ops[39-dtype0]" time="0.003"></testcase><testcase classname="tests.test_lorentz_math" file="tests/test_lorentz_math.py" line="209" name="test_parallel_transport_a_b[39-dtype0]" time="0.011"></testcase></testsuite>
```
Comment (ferrine) [inline test-output.xml]: oh, this was added by mistake

**Comment by ferrine** [comment]
I'll test against next pytorch soon too

File: test-output.xml (hunk 18 above)
Comment (rrkarim) [reply test-output.xml]: bit this outdated
